### DISCUSSION TITLE
Possible to nest a patrol test inside another patrol test?

### DISCUSSION THREAD

**Discussion by EyedBread** [leancodepl/patrol#1458] [Q&A]
When creating integration tests of our app, every test will need to do the same redundant task of logging in to the app which will result in a lot of code duplication. Is there a way for a patrolTest function to call upon another function which calls a patrolTest function? Like this:
```
void main() {
  patrolTest(
    'Load the app and login using test account and do something else',
    nativeAutomation: true,
    (PatrolTester $) async {

      await $.pumpWidgetAndSettle(exampleApp);

      await testLogin($);
      
      //Test some feature now while logged in to the app
      
      return;
    },
  );
}

Future<void> testLogin(PatrolTester $) async {
  patrolTest(
    "Login test",
    nativeAutomation: true,
    ($) async {
      try {
        await $('Next').tap();
        await $('Next').tap();
        await $('Explore').tap();
        await $('Accept Terms of Service').tap();
        await $('Sign in').tap();
        await $.native.selectFineLocation();
        await $.native.grantPermissionWhenInUse();
        await $(#PhoneNumberTextField).waitUntilVisible().enterText(
            '123456',
            visibleTimeout: const Duration(seconds: 15));
        await $('Sign in').tap();
      } catch (err) {
        //Log something
      }
    },
  );
}

```

The code above doesn't seem to work. How could I do this?

**Comment by bartekpacia** [comment]
In general, tests should not depend on one another. It's a bad thing. For this reason (and a few and other), you cannot nest patrolTest() in one another.

What I suggest is to extract the common sign-in logic to a function. Then you'll be able to simply call that function at the beginning of a test.

There's also another solution: you can pump your app in a state with a dummy user already signed in. That's more involved and easier to get wrong, but it's definitely possible, and might make sense if the sign-in process is long. But bear in mind that in this case your test won't be truly end-to-end.
